File: python/Arrays&Hashing/contains_duplicate.py
from typing import List

class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
        """
        Approach:
        - Use a hash set to track seen numbers.
        - If a number is already in the set, return True (duplicate found).
        - Otherwise, add the number to the set.
        - If no duplicates are found, return False.

        Time Complexity: O(n), where n is the length of nums (set operations are O(1) on average).
        Space Complexity: O(n), for the hash set.
        """
        hashset = set()
        for n in nums:
            if n in hashset:
                return True
            hashset.add(n)
        return False

class Solution2:
    def hasDuplicate(self, nums: List[int]) -> bool:
        """
        Approach:
        - Sort the array and check for consecutive duplicates.
        - If any two consecutive numbers are equal, return True.
        - Otherwise, return False.

        Time Complexity: O(n log n), due to sorting.
        Space Complexity: O(1) if sorting in place, O(n) if not.
        """
        nums.sort()
        i = 0
        while i < len(nums) - 1:
            if nums[i] == nums[i + 1]:
                return True
            i += 1
        return False

# -------------------
# Test Cases
# -------------------
def test_contains_duplicate():
    s1 = Solution()
    s2 = Solution2()
    # Test case 1: Contains duplicate
    nums1 = [1, 2, 3, 1]
    assert s1.hasDuplicate(nums1) == True
    assert s2.hasDuplicate(nums1) == True
    # Test case 2: No duplicate
    nums2 = [1, 2, 3, 4]
    assert s1.hasDuplicate(nums2) == False
    assert s2.hasDuplicate(nums2) == False
    # Test case 3: Empty list
    nums3 = []
    assert s1.hasDuplicate(nums3) == False
    assert s2.hasDuplicate(nums3) == False
    # Test case 4: Single element
    nums4 = [42]
    assert s1.hasDuplicate(nums4) == False
    assert s2.hasDuplicate(nums4) == False
    # Test case 5: All duplicates
    nums5 = [7, 7, 7, 7]
    assert s1.hasDuplicate(nums5) == True
    assert s2.hasDuplicate(nums5) == True
    print("All test cases passed!")

if __name__ == "__main__":
    test_contains_duplicate()

File: C++/Arrays&Hashing/contains_duplicate.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Approach 1: HashSet Solution
     * - Use HashSet to track seen numbers
     * - For each number, check if already in set
     * - If found, return true (duplicate exists)
     * - If not found, add to set
     * 
     * Time Complexity: O(n)
     * - Single pass through array
     * - HashSet operations (insert/lookup) are O(1) average
     * 
     * Space Complexity: O(n)
     * - HashSet can store up to n elements
     */
    bool hasDuplicate(vector<int>& nums) {
        unordered_set<int> hashSet;
        for (int n : nums){
            if (hashSet.count(n) > 0){
                return true;
            }
            else{
                hashSet.insert(n);
            }
        }
        return false;
    }
};

class Solution2 {
public:
    /*
     * Approach 2: Sorting Solution
     * - Sort array first
     * - Check adjacent elements for duplicates
     * - If adjacent elements match, found duplicate
     * 
     * Time Complexity: O(n log n)
     * - Dominated by sorting operation
     * - Linear scan after sorting is O(n)
     * 
     * Space Complexity: O(1)
     * - Sort is in-place
     * - Only uses constant extra space
     */
    bool hasDuplicate(vector<int>& nums) {
        if (nums.empty()) return false;
        
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size() - 1; i++) {
            if (nums[i] == nums[i + 1]) {
                return true;
            }
        }
        return false;
    }
};

/*
 * Detailed Syntax Explanation:
 * 
 * 1. Function Parameters:
 *    bool hasDuplicate(vector<int>& nums)
 *    - Returns boolean (true/false)
 *    - Takes reference to vector of integers
 *    - & prevents copying of vector
 *
 * 2. HashSet Solution:
 *    unordered_set<int> hashSet;
 *    - Creates empty HashSet for integers
 *    - O(1) average time for operations
 *
 *    for (int n : nums)
 *    - Range-based for loop
 *    - Iterates through each element
 *
 *    hashSet.count(n)
 *    - Returns number of occurrences (0 or 1)
 *
 * 3. Sorting Solution:
 *    if (nums.empty())
 *    - Checks for empty vector
 *    - Early return optimization
 *
 *    sort(nums.begin(), nums.end())
 *    - STL sort function
 *    - Takes iterator range
 *    - Modifies vector in-place
 *
 *    nums.size() - 1
 *    - Prevents buffer overflow
 *    - Allows checking adjacent elements
 */